File: interfaces/web/static/core/event-bus.js
// core/event-bus.js - EventSource client for real-time server events
import { sessionId } from '../shared/fetch.js';

let eventSource = null;
let reconnectAttempts = 0;
let reconnectTimeout = null;
let knownBootVersion = null;
const MAX_RECONNECT_DELAY = 30000;
const BASE_RECONNECT_DELAY = 1000;

// Event handlers registry
const handlers = new Map();

/**
 * Connect to the event bus SSE endpoint
 * @param {boolean} replay - Whether to replay recent events on connect
 */
export function connect(replay = false) {
    if (eventSource && eventSource.readyState !== EventSource.CLOSED) {
        console.log('[EventBus] Already connected');
        return;
    }
    
    const url = `/api/events?replay=${replay}`;
    console.log(`[EventBus] Connecting to ${url}`);
    
    eventSource = new EventSource(url);
    
    eventSource.onopen = () => {
        console.log('[EventBus] Connected');
        reconnectAttempts = 0;
        dispatch('bus_connected', {});
    };
    
    eventSource.onmessage = (e) => {
        try {
            const event = JSON.parse(e.data);

            // Skip keepalives
            if (event.type === 'keepalive') return;

            // Detect server restart via boot_version change
            if (event.type === 'connected' && event.data?.boot_version) {
                const v = event.data.boot_version;
                if (knownBootVersion && knownBootVersion !== v) {
                    console.log(`[EventBus] Server restarted (boot ${knownBootVersion} → ${v})`);
                    dispatch('server_restarted', { old: knownBootVersion, new: v });
                }
                knownBootVersion = v;
            }

            // Skip events we originated (we already handled them locally)
            if (event.data?.origin && event.data.origin === sessionId) {
                console.log(`[EventBus] Skipping self-originated ${event.type}`);
                return;
            }

            // Dispatch to registered handlers
            dispatch(event.type, event.data, event.timestamp);

        } catch (err) {
            console.error('[EventBus] Parse error:', err, e.data);
        }
    };
    
    eventSource.onerror = (e) => {
        console.warn('[EventBus] Connection error, will reconnect');
        eventSource.close();
        scheduleReconnect();
    };
}

/**
 * Disconnect from the event bus
 */
export function disconnect() {
    if (reconnectTimeout) {
        clearTimeout(reconnectTimeout);
        reconnectTimeout = null;
    }
    
    if (eventSource) {
        eventSource.close();
        eventSource = null;
        console.log('[EventBus] Disconnected');
        dispatch('bus_disconnected', {});
    }
}

/**
 * Schedule a reconnection with exponential backoff
 */
function scheduleReconnect() {
    if (reconnectTimeout) return;
    
    const delay = Math.min(
        BASE_RECONNECT_DELAY * Math.pow(2, reconnectAttempts),
        MAX_RECONNECT_DELAY
    );
    
    console.log(`[EventBus] Reconnecting in ${delay}ms (attempt ${reconnectAttempts + 1})`);
    
    reconnectTimeout = setTimeout(() => {
        reconnectTimeout = null;
        reconnectAttempts++;
        connect(false); // Don't replay on reconnect
    }, delay);
}

/**
 * Register an event handler
 * @param {string} eventType - Event type to listen for
 * @param {function} handler - Handler function(data, timestamp)
 * @returns {function} Unsubscribe function
 */
export function on(eventType, handler) {
    if (!handlers.has(eventType)) {
        handlers.set(eventType, new Set());
    }
    handlers.get(eventType).add(handler);
    
    // Return unsubscribe function
    return () => {
        const set = handlers.get(eventType);
        if (set) {
            set.delete(handler);
            if (set.size === 0) {
                handlers.delete(eventType);
            }
        }
    };
}

/**
 * Remove an event handler
 * @param {string} eventType - Event type
 * @param {function} handler - Handler to remove
 */
export function off(eventType, handler) {
    const set = handlers.get(eventType);
    if (set) {
        set.delete(handler);
    }
}

/**
 * Dispatch an event to all registered handlers
 */
export function dispatch(eventType, data, timestamp) {
    const set = handlers.get(eventType);
    if (set) {
        for (const handler of set) {
            try {
                handler(data, timestamp);
            } catch (err) {
                console.error(`[EventBus] Handler error for ${eventType}:`, err);
            }
        }
    }
    
    // Also dispatch wildcard handlers
    const wildcardSet = handlers.get('*');
    if (wildcardSet) {
        for (const handler of wildcardSet) {
            try {
                handler({ type: eventType, data, timestamp });
            } catch (err) {
                console.error('[EventBus] Wildcard handler error:', err);
            }
        }
    }
}

/**
 * Check if connected
 * @returns {boolean}
 */
export function isConnected() {
    return eventSource && eventSource.readyState === EventSource.OPEN;
}

// Clean up on page unload to prevent stale connections
window.addEventListener('beforeunload', () => {
    if (eventSource) {
        eventSource.close();
        eventSource = null;
    }
});

// Event type constants (mirror server-side)
export const Events = {
    // AI/Chat events
    AI_TYPING_START: 'ai_typing_start',
    AI_TYPING_END: 'ai_typing_end',
    MESSAGE_ADDED: 'message_added',
    MESSAGE_REMOVED: 'message_removed',
    CHAT_SWITCHED: 'chat_switched',
    CHAT_CREATED: 'chat_created',
    CHAT_CLEARED: 'chat_cleared',
    
    // TTS events
    TTS_PLAYING: 'tts_playing',
    TTS_STOPPED: 'tts_stopped',
    TTS_SPEAK: 'tts_speak',
    
    // STT events
    STT_RECORDING_START: 'stt_recording_start',
    STT_RECORDING_END: 'stt_recording_end',
    STT_PROCESSING: 'stt_processing',
    
    // Wakeword events
    WAKEWORD_DETECTED: 'wakeword_detected',
    
    // Tool events
    TOOL_EXECUTING: 'tool_executing',
    TOOL_COMPLETE: 'tool_complete',
    
    // System events
    PROMPT_CHANGED: 'prompt_changed',
    PROMPT_DELETED: 'prompt_deleted',
    COMPONENTS_CHANGED: 'components_changed',
    TOOLSET_CHANGED: 'toolset_changed',
    SPICE_CHANGED: 'spice_changed',
    SETTINGS_CHANGED: 'settings_changed',
    CHAT_SETTINGS_CHANGED: 'chat_settings_changed',
    
    // Context threshold events
    CONTEXT_WARNING: 'context_warning',
    CONTEXT_CRITICAL: 'context_critical',
    
    // Error events
    LLM_ERROR: 'llm_error',
    TTS_ERROR: 'tts_error',
    STT_ERROR: 'stt_error',
    
    // Continuity events
    CONTINUITY_TASK_STARTING: 'continuity_task_starting',
    CONTINUITY_TASK_COMPLETE: 'continuity_task_complete',
    CONTINUITY_TASK_PROGRESS: 'continuity_task_progress',
    CONTINUITY_TASK_ERROR: 'continuity_task_error',

    // Plugin events
    PLUGIN_RELOADED: 'plugin_reloaded',
    PLUGIN_LOAD_ERROR: 'plugin_load_error',
    PLUGIN_NOTICE: 'plugin_notice',   // generic plugin->UI toast: {plugin, message, severity}

    // Mind data changed — tool or UI wrote to memory/goal/knowledge/people.
    // Payload: {domain, scope, action}. Mind view uses it to live-refresh.
    MIND_CHANGED: 'mind_changed',

    // Re-embed pipeline progress. Payload = full status snapshot from the
    // worker (running, total, done, current_table, errors, last_error).
    REEMBED_PROGRESS: 'reembed_progress',

    // Agent events
    AGENT_SPAWNED: 'agent_spawned',
    AGENT_COMPLETED: 'agent_completed',
    AGENT_DISMISSED: 'agent_dismissed',

    // User interaction events (avatar state triggers + overlay)
    USER_TYPING: 'user_typing',       // User is typing in chat input
    USER_SENT: 'user_sent',           // User sent a message — data: {text}
    CHAT_CHUNK: 'chat_chunk',         // AI streaming text chunk — data: {text}

    // Connection events
    CONNECTED: 'connected',           // Server confirms SSE subscription
    BUS_CONNECTED: 'bus_connected',   // Client-side: EventSource opened
    BUS_DISCONNECTED: 'bus_disconnected',
    SERVER_RESTARTED: 'server_restarted'  // Boot version changed — caches stale
};

File: core/event_bus.py
# event_bus.py - Central pub/sub event bus for real-time UI updates
import asyncio
import threading
import queue
import time
import json
import logging
from typing import AsyncGenerator, Generator, Optional, Dict, Any
from collections import deque

logger = logging.getLogger(__name__)

class EventBus:
    """Thread-safe pub/sub event bus with replay buffer for late subscribers."""
    
    def __init__(self, replay_size: int = 50):
        self._lock = threading.Lock()
        self._subscribers: Dict[str, queue.Queue] = {}
        self._async_subscribers: Dict[str, tuple] = {}  # sub_id -> (asyncio.Queue, loop)
        self._replay_buffer: deque = deque(maxlen=replay_size)
        self._subscriber_counter = 0
        logger.info(f"EventBus initialized (replay_size={replay_size})")
    
    def publish(self, event_type: str, data: Optional[Dict[str, Any]] = None):
        """Publish an event to all subscribers (sync and async)."""
        event = {
            "type": event_type,
            "data": data or {},
            "timestamp": time.time()
        }

        with self._lock:
            # Ephemeral events (transient UI toasts) aren't replayed to late
            # subscribers - a freshly-opened tab shouldn't surface a stale "done" toast.
            if event_type not in ("plugin_notice",):
                self._replay_buffer.append(event)
            dead_subscribers = []

            # Sync subscribers
            for sub_id, q in self._subscribers.items():
                try:
                    q.put_nowait(event)
                except queue.Full:
                    logger.warning(f"Subscriber {sub_id} queue full, dropping event")
                except Exception as e:
                    logger.error(f"Error publishing to {sub_id}: {e}")
                    dead_subscribers.append(sub_id)

            for sub_id in dead_subscribers:
                del self._subscribers[sub_id]

            # Async subscribers — thread-safe put via event loop
            dead_async = []
            for sub_id, (aq, loop) in self._async_subscribers.items():
                try:
                    loop.call_soon_threadsafe(aq.put_nowait, event)
                except RuntimeError:
                    dead_async.append(sub_id)

            for sub_id in dead_async:
                del self._async_subscribers[sub_id]

        logger.debug(f"Published: {event_type}")
    
    def subscribe(self, replay: bool = True) -> Generator[Dict[str, Any], None, None]:
        """Subscribe to events. Yields events as they arrive.
        
        Args:
            replay: If True, replay recent events before live stream
        """
        sub_id = None
        q = queue.Queue(maxsize=100)
        
        with self._lock:
            self._subscriber_counter += 1
            sub_id = f"sub_{self._subscriber_counter}"
            self._subscribers[sub_id] = q
            
            if replay:
                for event in self._replay_buffer:
                    try:
                        q.put_nowait(event)
                    except queue.Full:
                        break
        
        logger.info(f"New subscriber: {sub_id} (replay={replay}) — total subscribers: {len(self._subscribers)}")

        # Immediate connection event - wakes up client instantly
        # Include boot_version so frontend can detect server restarts
        boot_version = None
        try:
            from core.api_fastapi import BOOT_VERSION
            boot_version = BOOT_VERSION
        except Exception:
            pass
        yield {"type": "connected", "data": {"sub_id": sub_id, "boot_version": boot_version}, "timestamp": time.time()}

        try:
            keepalive_count = 0
            while True:
                try:
                    event = q.get(timeout=15)
                    yield event
                except queue.Empty:
                    # Send keepalive (15s interval)
                    keepalive_count += 1
                    logger.debug(f"Keepalive #{keepalive_count} for {sub_id}")
                    yield {"type": "keepalive", "timestamp": time.time()}
        except GeneratorExit:
            logger.info(f"Subscriber {sub_id} generator closed by client")
        finally:
            with self._lock:
                if sub_id in self._subscribers:
                    del self._subscribers[sub_id]
            logger.info(f"Subscriber disconnected: {sub_id}")
    
    async def async_subscribe(self, replay: bool = True) -> AsyncGenerator[Dict[str, Any], None]:
        """Async subscribe to events. No threadpool thread consumed."""
        sub_id = None
        aq = asyncio.Queue(maxsize=100)
        loop = asyncio.get_running_loop()

        with self._lock:
            self._subscriber_counter += 1
            sub_id = f"async_sub_{self._subscriber_counter}"
            self._async_subscribers[sub_id] = (aq, loop)

            if replay:
                for event in self._replay_buffer:
                    try:
                        aq.put_nowait(event)
                    except asyncio.QueueFull:
                        break

        logger.info(f"New async subscriber: {sub_id} (replay={replay}) — total: {len(self._subscribers) + len(self._async_subscribers)}")

        boot_version = None
        try:
            from core.api_fastapi import BOOT_VERSION
            boot_version = BOOT_VERSION
        except Exception:
            pass
        yield {"type": "connected", "data": {"sub_id": sub_id, "boot_version": boot_version}, "timestamp": time.time()}

        try:
            keepalive_count = 0
            while True:
                try:
                    event = await asyncio.wait_for(aq.get(), timeout=15)
                    yield event
                except asyncio.TimeoutError:
                    keepalive_count += 1
                    logger.debug(f"Keepalive #{keepalive_count} for {sub_id}")
                    yield {"type": "keepalive", "timestamp": time.time()}
        except GeneratorExit:
            logger.info(f"Async subscriber {sub_id} generator closed by client")
        finally:
            with self._lock:
                self._async_subscribers.pop(sub_id, None)
            logger.info(f"Async subscriber disconnected: {sub_id}")

    def subscriber_count(self) -> int:
        """Return current number of subscribers."""
        with self._lock:
            return len(self._subscribers) + len(self._async_subscribers)


# Singleton instance
_bus: Optional[EventBus] = None

def get_event_bus() -> EventBus:
    """Get or create the singleton event bus."""
    global _bus
    if _bus is None:
        _bus = EventBus()
    return _bus

def publish(event_type: str, data: Optional[Dict[str, Any]] = None):
    """Convenience function to publish to the global bus."""
    get_event_bus().publish(event_type, data)


# Event type constants
class Events:
    # AI/Chat events
    AI_TYPING_START = "ai_typing_start"
    AI_TYPING_END = "ai_typing_end"
    MESSAGE_ADDED = "message_added"
    MESSAGE_REMOVED = "message_removed"
    CHAT_SWITCHED = "chat_switched"
    CHAT_CREATED = "chat_created"
    CHAT_CLEARED = "chat_cleared"
    
    # TTS events
    TTS_PLAYING = "tts_playing"
    TTS_STOPPED = "tts_stopped"
    TTS_SPEAK = "tts_speak"
    
    # STT events
    STT_RECORDING_START = "stt_recording_start"
    STT_RECORDING_END = "stt_recording_end"
    STT_PROCESSING = "stt_processing"
    
    # Wakeword events
    WAKEWORD_DETECTED = "wakeword_detected"

    # Conversation mode (v3) — fired on enter/exit so the UI reflects state.
    # Payload: {enabled: bool}.
    CONVERSATION_MODE_CHANGED = "conversation_mode_changed"
    
    # Tool events
    TOOL_EXECUTING = "tool_executing"
    TOOL_COMPLETE = "tool_complete"
    
    # System events
    PROMPT_CHANGED = "prompt_changed"
    PROMPT_DELETED = "prompt_deleted"
    COMPONENTS_CHANGED = "components_changed"
    TOOLSET_CHANGED = "toolset_changed"
    SPICE_CHANGED = "spice_changed"
    BACKGROUND_CHANGED = "background_changed"            # active chat background changed (render)
    BACKGROUNDS_LIBRARY_CHANGED = "backgrounds_library_changed"  # library upload/delete (tool refresh)
    SETTINGS_CHANGED = "settings_changed"
    CHAT_SETTINGS_CHANGED = "chat_settings_changed"
    # Scope CRUD — fired when memory/knowledge/goal/people scopes are created or
    # deleted, so the chat sidebar can refresh its dropdowns without a page
    # reload. Payload: {kind: "memory"|"knowledge"|"goal"|"people", action:
    # "created"|"deleted", name: "<scope_name>"}.
    SCOPE_CHANGED = "scope_changed"
    
    # Context threshold events
    CONTEXT_WARNING = "context_warning"    # 80% threshold
    CONTEXT_CRITICAL = "context_critical"  # 95% threshold
    
    # Connection events
    CONNECTED = "connected"

    # Error events
    LLM_ERROR = "llm_error"
    TTS_ERROR = "tts_error"
    STT_ERROR = "stt_error"
    
    # Continuity events
    CONTINUITY_TASK_STARTING = "continuity_task_starting"
    CONTINUITY_TASK_COMPLETE = "continuity_task_complete"
    CONTINUITY_TASK_SKIPPED = "continuity_task_skipped"
    CONTINUITY_TASK_ERROR = "continuity_task_error"
    CONTINUITY_TASK_PROGRESS = "continuity_task_progress"

    # Plugin events
    PLUGIN_RELOADED = "plugin_reloaded"
    PLUGIN_LOAD_ERROR = "plugin_load_error"
    # Generic plugin -> UI toast. Any plugin may publish this to surface a toast.
    # Payload: {plugin, message, severity} where severity in
    # {info, success, warning, error}. (Added 2026-06-16.)
    PLUGIN_NOTICE = "plugin_notice"

    # Daemon/webhook events
    DAEMON_EVENT = "daemon_event"
    WEBHOOK_FIRED = "webhook_fired"

    # Mind-data changed events — fired when the AI writes to memory/goal/
    # knowledge/people stores so the Mind view can live-refresh instead of
    # showing stale content after a save. Payload: {domain, scope, action}
    # where domain in {memory, goal, knowledge, people}, action in
    # {save, update, delete}. Without this event the user sees "done"
    # from Sapphire but their Mind tab doesn't update — can't tell the
    # difference between "tool silently failed" and "tool worked, view
    # stale." AIX-class bug. (Added 2026-04-19 after Scout 3 dispatch.)
    MIND_CHANGED = "mind_changed"

    # Re-embed pipeline progress — fired by the background worker as it
    # walks memories/knowledge_entries/people, re-generating vectors under
    # the current provider. Payload is the full status snapshot (running,
    # total, done, current_table, errors, etc). Settings UI uses this to
    # show a live progress bar without polling.
    REEMBED_PROGRESS = "reembed_progress"

    # Agent events
    AGENT_SPAWNED = "agent_spawned"
    AGENT_COMPLETED = "agent_completed"
    AGENT_DISMISSED = "agent_dismissed"
    AGENT_BATCH_COMPLETE = "agent_batch_complete"
    WORKSPACE_READY = "workspace_ready"